metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala

```diff
@@ -0,0 +1,277 @@
+package scala.meta.internal.metals
+
+import java.io.ByteArrayOutputStream
+import java.io.PrintStream
+import java.net.URI
+import java.net.URLDecoder
+import java.nio.charset.StandardCharsets
+import java.nio.file.Paths
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.util.Properties
+import scala.util.Try
+
+import scala.meta.cli.Reporter
+import scala.meta.internal.builds.ShellRunner
+import scala.meta.internal.io.FileIO
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metap.Main
+import scala.meta.internal.mtags.SemanticdbClasspath
+import scala.meta.io.AbsolutePath
+import scala.meta.metap.Format
+import scala.meta.metap.Settings
+
+final class FileDecoderProvider(
+    workspace: AbsolutePath,
+    buildTargets: BuildTargets,
+    userConfig: () => UserConfiguration,
+    shellRunner: ShellRunner,
+    fileSystemSemanticdbs: FileSystemSemanticdbs
+)(implicit ec: ExecutionContext) {
+
+  private case class Finder(findFile: String => Option[AbsolutePath])
+  private case class Decoder(decode: AbsolutePath => Future[Option[String]])
+
+  /**
+   * URI format...
+   * metalsDecodeFile:/fileUrl?decoder=decoderName&keyValueArguments
+   * jar:file:///jarPath/jar-sources.jar!/packagedir/file.java
+   *
+   * Examples...
+   * metalsDecodeFile:/somedir/someFile.java.javap.verbose?decoder=javap
+   * metalsDecodeFile:/somedir/someFile.class.javap?decoder=javap&verbose=false
```

Comment:
I'm not fully sure what's happening with these, but I'm unable to get this to work. I keep getting:

```
metalsDecodeFile:/Users/ckipp/Documents/sanity/src/main/scala/Main.java.javap.verbose?decoder=javap
Unexpected uri, Metals accepts ones such as:
metalsDecodeFile:/somedir/someFile.java.javap.verbose?decoder=javap
Take a look at scala/meta/internal/metals/FileDecoderProvider for more information.
```
Just to verify, how come when we do semanticdb we do `someFile.scala?decoder=semanticdb` and then here we do `someFile.java.javap?decoder=javap`. I would have assumed this should have been `someFile.scala?decoder-javap`.

Reply:
The URIs are from me mucking about with Java files.

If you have a scala source then use `someFile.scala.javap`

It's also possible to run semanticdb and javap decoders directly on the semanticdb and class file URIs in the `.bloop/**` dirs

Reply:
🤔 So I assumed that it was `.scala` at first and no matter what I try I keep getting:

```
Error: class not found: Simple.class
```

No matter what file I'm trying it in. What I end up sending in is:

```
[Trace - 08:37:51 AM] Received request 'workspace/executeCommand - (13)'
Params: {
  "command": "metals.file-decode",
  "arguments": [
    "metalsDecodeFile:///Users/ckipp/Documents/sanity/src/main/scala/Simple.scala.javap?decoder\u003djavap\u0026verbose\u003dfalse"
  ]
}
```

Which should be fine no?

Reply:
It should be.  Is there a class file present? I imagine there should be if the semanticdb one is present.

Something like: `./bloop/some bsp classes path/sanity/Simple.class`?

You should be able to run javap directly on that class file as well as the source file e.g.
`metalsDecodeFile:///Users/ckipp/Documents/sanity/.bloop/XXXX/Simple.class.javap?decoder\u003djavap\u0026verbose\u003dfalse`
If that works then for some reason it's not getting the build target correctly and then working out where the class file should be

I'm surprised you're getting that error - I thought all errors were intercepted to just return invalid URI info.

Reply:
Yea it's odd. It works for me when I view semanticdb, but not javap.

![2021-09-14 14 07 22](https://user-images.githubusercontent.com/13974112/133254849-255a8eb5-b31c-4c53-8565-fb528117a841.gif)

> You should be able to run javap directly on that class file as well as the source file e.g.
metalsDecodeFile:///Users/ckipp/Documents/sanity/.bloop/XXXX/Simple.class.javap?decoder\u003djavap\u0026verbose\u003dfalse
If that works then for some reason it's not getting the build target correctly and then working out where the class file should be

This is much trickier in other clients, and I haven't actually looked into this yet. `nvim-metals` works by detecting the file type and then deciding whether or not it should "attach" the LSP server to that buffer. So in this situation `.class` files aren't something that are recognized, so if you open them Metals isn't attatched. It's something I'll need to figure out.

Reply:
Ah - I think this is actually the output of `javap`.  can you try javap from the CLI on that class file?

Reply:
<img width="998" alt="Screenshot 2021-09-14 at 14 24 51" src="https://user-images.githubusercontent.com/13974112/133257028-3e6a2417-cad7-4cbf-800c-885564b7fdbd.png">

Works fine from the CLI

Reply:
It may be that I need to specify the classpath in the CLI even though it runs from that folder.
Could you try `javap -cp . -verbose Simple.class`

Reply:
> It may be that I need to specify the classpath in the CLI even though it runs from that folder.
Could you try javap -cp . -verbose Simple.class

This shows everything

```
~/Documents/scala-workspace/sanity/.bloop/foo/bloop-bsp-clients-classes/classes-Metals-oOUyR9noRRmJGfmsWpXaXg==/sanity
❯ javap -cp . -verbose Simple.class
Classfile /Users/ckipp/Documents/scala-workspace/sanity/.bloop/foo/bloop-bsp-clients-classes/classes-Metals-oOUyR9noRRmJGfmsWpXaXg==/sanity/Simple.class
  Last modified 14 Sep 2021; size 599 bytes
  MD5 checksum 270ecda1de731925fae6153cc6190c3e
  Compiled from "Simple.scala"
public final class sanity.Simple
  minor version: 0
  major version: 52
  flags: (0x0031) ACC_PUBLIC, ACC_FINAL, ACC_SUPER
  this_class: #2                          // sanity/Simple
  super_class: #4                         // java/lang/Object
  interfaces: 0, fields: 0, methods: 1, attributes: 3
Constant pool:
   #1 = Utf8               sanity/Simple
   #2 = Class              #1             // sanity/Simple
   #3 = Utf8               java/lang/Object
   #4 = Class              #3             // java/lang/Object
   #5 = Utf8               Simple.scala
   #6 = Utf8               Lscala/reflect/ScalaSignature;
   #7 = Utf8               bytes
   #8 = Utf8               \u0006\u0001y9Q!\u0002\u0004\t\u0002%1Qa\u0003\u0004\t\u00021AQaE\u0001\u0005\u0002QAq!F\u0001C\u0002\u0013\u0005a\u0003\u0003\u0004\u001e\u0003\u0001\u0006IaF\u0001\u0007\'&l\u0007\u000f\\3\u000b\u0003\u001d\taa]1oSRL8\u0001\u0001\t\u0003\u0015\u0005i\u0011A\u0002\u0002\u0007\'&l\u0007\u000f\\3\u0014\u0005\u0005i\u0001C\u0001\b\u0012\u001b\u0005y!\"\u0001\t\u0002\u000bM\u001c\u0017\r\\1\n\u0005Iy!AB!osJ+g-\u0001\u0004=S:LGO\u0010\u000b\u0002\u0013\u0005)Ao\u001c;bYV\tq\u0003E\u0002\u000f1iI!!G\b\u0003\r=\u0003H/[8o!\tq1$\u0003\u0002\u001d\u001f\t\u0019\u0011J\u001c;\u0002\rQ|G/\u00197!\u0001
   #9 = Utf8               total
  #10 = Utf8               ()Lscala/Option;
  #11 = Utf8               ()Lscala/Option<Ljava/lang/Object;>;
  #12 = Utf8               sanity/Simple$
  #13 = Class              #12            // sanity/Simple$
  #14 = Utf8               MODULE$
  #15 = Utf8               Lsanity/Simple$;
  #16 = NameAndType        #14:#15        // MODULE$:Lsanity/Simple$;
  #17 = Fieldref           #13.#16        // sanity/Simple$.MODULE$:Lsanity/Simple$;
  #18 = NameAndType        #9:#10         // total:()Lscala/Option;
  #19 = Methodref          #13.#18        // sanity/Simple$.total:()Lscala/Option;
  #20 = Utf8               Code
  #21 = Utf8               Signature
  #22 = Utf8               SourceFile
  #23 = Utf8               RuntimeVisibleAnnotations
  #24 = Utf8               ScalaSig
{
  public static scala.Option<java.lang.Object> total();
    descriptor: ()Lscala/Option;
    flags: (0x0009) ACC_PUBLIC, ACC_STATIC
    Code:
      stack=1, locals=0, args_size=0
         0: getstatic     #17                 // Field sanity/Simple$.MODULE$:Lsanity/Simple$;
         3: invokevirtual #19                 // Method sanity/Simple$.total:()Lscala/Option;
         6: areturn
    Signature: #11                          // ()Lscala/Option<Ljava/lang/Object;>;
}
SourceFile: "Simple.scala"
RuntimeVisibleAnnotations:
  0: #6(#7=s#8)
    scala.reflect.ScalaSignature(
      bytes="\u0006\u0001y9Q!\u0002\u0004\t\u0002%1Qa\u0003\u0004\t\u00021AQaE\u0001\u0005\u0002QAq!F\u0001C\u0002\u0013\u0005a\u0003\u0003\u0004\u001e\u0003\u0001\u0006IaF\u0001\u0007\'&l\u0007\u000f\\3\u000b\u0003\u001d\taa]1oSRL8\u0001\u0001\t\u0003\u0015\u0005i\u0011A\u0002\u0002\u0007\'&l\u0007\u000f\\3\u0014\u0005\u0005i\u0001C\u0001\b\u0012\u001b\u0005y!\"\u0001\t\u0002\u000bM\u001c\u0017\r\\1\n\u0005Iy!AB!osJ+g-\u0001\u0004=S:LGO\u0010\u000b\u0002\u0013\u0005)Ao\u001c;bYV\tq\u0003E\u0002\u000f1iI!!G\b\u0003\r=\u0003H/[8o!\tq1$\u0003\u0002\u001d\u001f\t\u0019\u0011J\u001c;\u0002\rQ|G/\u00197!\u0001"
    )
  ScalaSig: length = 0x3 (unknown attribute)
   05 00 00
``

Reply:
I'm pretty sure the `Error: class not found: Main.class` is the output of `javap`.

Could you log what is going on in `FileDecoderProvider#decodeJavapFromClassFile`
Specifically what the `filePath.filename` and `filePath.parent` is.
I wonder if this is a symlink thing.  I'm on Windows + WSL2 but maybe it's not handling something on the Mac